devtools/maven/src/main/java/io/quarkus/maven/DeployMojo.java

```diff
@@ -1,21 +1,77 @@
 import org.apache.maven.plugins.annotations.ResolutionScope;
 
+import io.quarkus.bootstrap.app.AugmentAction;
+import io.quarkus.bootstrap.app.CuratedApplication;
+import io.quarkus.deployment.cmd.DeployCommandActionResultBuildItem;
+import io.quarkus.deployment.cmd.DeployCommandDeclarationHandler;
+import io.quarkus.deployment.cmd.DeployCommandDeclarationResultBuildItem;
+import io.quarkus.deployment.cmd.DeployCommandHandler;
+
 @Mojo(name = "deploy", defaultPhase = LifecyclePhase.PACKAGE, requiresDependencyResolution = ResolutionScope.COMPILE_PLUS_RUNTIME, threadSafe = true)
 public class DeployMojo extends AbstractDeploymentMojo {
 
     @Override
     protected boolean beforeExecute() throws MojoExecutionException {
-        systemProperties = new HashMap<>(systemProperties);
-        boolean shouldBuildImage = imageBuild || imageBuilder != null && !imageBuilder.isEmpty();
-        systemProperties.put("quarkus." + getDeployer().name() + ".deploy", "true");
-        systemProperties.put("quarkus.container-image.build", String.valueOf(shouldBuildImage));
         return super.beforeExecute();
     }
+
+    @Override
+    protected void doExecute() throws MojoExecutionException {
+        try (CuratedApplication curatedApplication = bootstrapApplication()) {
+            // get list of extensions that support quarkus:deploy
+            AtomicReference<List<String>> tooMany = new AtomicReference<>();
+            AugmentAction action = curatedApplication.createAugmentor();
+            action.performCustomBuild(DeployCommandDeclarationHandler.class.getName(), new Consumer<List<String>>() {
+                @Override
+                public void accept(List<String> strings) {
+                    tooMany.set(strings);
+                }
+            }, DeployCommandDeclarationResultBuildItem.class.getName());
+            String target = System.getProperty("quarkus.deploy.target");
+            List<String> targets = tooMany.get();
+            if (targets.isEmpty() && target == null) {
+                // weave in kubernetes as we have no deploy support from others
+                systemProperties = new HashMap<>(systemProperties);
+                boolean shouldBuildImage = imageBuild || imageBuilder != null && !imageBuilder.isEmpty();
+                systemProperties.put("quarkus." + getDeployer().name() + ".deploy", "true");
+                systemProperties.put("quarkus.container-image.build", String.valueOf(shouldBuildImage));
+                super.doExecute();
+            } else if (targets.size() > 1 && target == null) {
+                getLog().error(
+                        "Too many installed extensions support quarkus:deploy.  You must choose one by setting quarkus.deploy.target.");
+                getLog().error("Extensions: " + targets.stream().collect(Collectors.joining(" ")));
+            } else if (target != null && !targets.contains(target)) {
+                getLog().error(
```

Comment:
In this case, we would like a matching extension (if found) to be added as a `forcedDependency`.

For example in any quarkus project the following should always work:

```sh
mvn quarkus:deploy -Dquarkus.deploy.target=knative
mvn quarkus:deploy -Dquarkus.deploy.target=openshift
```
Besides the hard coding approach which is what was originally used, I can think of two approaches that could possibly work:

1. Express deployment capabilites in the extension metadata and perform a registry lookup.
2. Use a convention like adding `quarkus-${quarkus.deploy.target|` to forced dependencies.

**Option 1**: has the downsides that its slow.
**Option 2**: is not the most reliable (e.g. in the existing codebase knative would still not work as there is no `quarkus-knative` extension).

Reply:
Forcing dependencies, IMO, is a mediocre solution, at best, when using CLI, but I don't like it at all from the mvn/gradle command line.  Forcing dependencies doesn't even work with Gradle, correct?  For Lambda and Azure Functions there are packaging requirements for quarkus:run and quarkus:deploy to work, so, forcing dependencies makes no sense for those plugins.

Any specific hardcoding should be removed from the maven/gradle code base, and only CLI should be able to force dependencies, IMO.   I would even go as far to say that forced dependencies should be removed as a solution as it doesn't work for Gradle at all.   I don't think its a burden to force users to pull in the dependency.

Reply:
Also, the kube extensions were not meant to use the quarkus.deploy.target until they are refactored to use the new deployment mechanism.

Reply:
>  Forcing dependencies doesn't even work with Gradle, correct? 
> 
Forcing dependecnies from within a gradle task doesn't work. However, when using the CLI we can `force` dependencies for gradle project by generating a gralde init script. This is something that we already do and it works fine.

> Any specific hardcoding should be removed from the maven/gradle code base,
>
I agree that we can remove hardcoding from maven/gralde and what you've put there is a great step towards this goal.

> I don't think its a burden to force users to pull in the dependency.
>
No, it is not. But then again, the kubernetes bits do not affect runtime, so there is not reason for them to be added to the project. And given that different people within a team may be using different tools (especially for the container image bits and local kube development like quarkus-minikube and quarkus-kind) it can cause uneeded noise git.

Reply:
I don't want to refactor the Kube extensions within this PR please.   Can we just get this approved and merged?  Is it not backward compatible as-is right now?

Reply:
The changes do seem to be backwards compatible. 
We do however two ways of doing pretty much the same thing (and these are not compatible).
We have `quarkus.deploy.target=xxxx` (new) and `quarkus.xxxx.deploy=true` (existing).

What's the intention here? 
If we keep them both it's going to be confusing to users.
If we merge them into one, we need to do it in a compatible way.

Reply:
Ok, I'll refactor kube extensions a little to use the quarkus.deploy.target?  The goal of that switch was it would be settable in application.properties too.  This would break backward compatibility though in the kube extensions.

Reply:
I am not asking you to refactor the kube extensions. I am just trying to understand what you have in mind.

Reply:
@patriot1burke, @maxandersen : How would you like us to proceed? Given that the code doesn't break compatibility I am ok with this PR. 

We can deal with `quarkus.deploy.target=xxxx` vs `quarkus.xxxx.deploy=true` in a follow up as long as we agree that we will somehow retain a way of forcing dependencies. What do you think?

Reply:
Will try again tomorrow after my talk.

Reply:
Up to you on work.  I can make the change to quarkus.xxx.deploy -> quarkus.deploy.target if you want.  Forced dependencies should be done before lookup of deploy options too. (if it isn't already)

Reply:
Hmmm, I don't think that we can completely remove `quarkus.xxx.deploy` at this point. At best we could deprecate it.